resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/CircuitBreaker.java

```diff
@@ -983,7 +1020,102 @@ static CircuitBreaker of(String name, Supplier<CircuitBreakerConfig> circuitBrea
     }
 
     /**
-     * Creates a CircuitBreaker with a custom CircuitBreaker configuration.
+     * This class decorates future to add CircuitBreaking functionality around invocation.
+     * @param <T>
+     */
+    final class CircuitBreakerFuture<T> implements Future<T> {
+        private CircuitBreaker circuitBreaker;
+        private Future<T> future;
+        private long start;
+        private boolean isPermissionAcquired;
+
+        CircuitBreakerFuture(CircuitBreaker circuitBreaker, Future<T> future, long start, boolean isPermissionAcquired) {
+            Objects.requireNonNull(future, "Non null Future is required to decorate");
+            this.circuitBreaker = circuitBreaker;
+            this.future = future;
+            this.start = start;
+            this.isPermissionAcquired = isPermissionAcquired;
+        }
+
+        CircuitBreakerFuture(CircuitBreaker circuitBreaker, Future<T> future) {
+            Objects.requireNonNull(future, "Non null Future is required to decorate");
+            this.circuitBreaker = circuitBreaker;
+            this.future = future;
+            this.start = System.nanoTime();
+            this.isPermissionAcquired = false;
+        }
+
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning) {
+            return future.cancel(mayInterruptIfRunning);
+        }
+
+        @Override
+        public boolean isCancelled() {
+            return future.isCancelled();
+        }
+
+        @Override
+        public boolean isDone() {
+            return future.isDone();
+        }
+
+        @Override
+        public T get() throws InterruptedException, ExecutionException {
+            if (!isPermissionAcquired && !circuitBreaker.tryAcquirePermission()) {
```

Comment:
Why do we need to check the permission again? 
The permission was already checked even before the method which returns a Future was invoked.

Reply:
The decorator itself should not spawn a thread. The method which returns the future controls the thread management. Our decorator should just do the permission check and call onSuccess or onError.

Reply:
When precisely should the decorator call `onSuccess()` or `onError()`? The first time a success or failure is observed? That could be a much longer interval than the time that the underlying call took to return. Short of pushing the call onto the thread that's already supplying the value of the Future (and how would one do that?), I'm not sure what else can be done here.

Reply:
After above discussion, I am inching towards adding a `permissionEvaluated` check in `CircuitBreakerFuture` class which bypass the CircuitBreaking logic if evaluated already. Please confirm?

@Clgroft makes a good point here, evaluating `future.get` is at discretion of caller and it could be a much longer interval than the time that the underlying call took to return. Where as in `CompletableFuture` `whenComplete` is called as soon as underlying computation is return. This is downside of creating your own `Future` and decorating it with CircuitBreaking logic. Let us know your thoughts.

Reply:
@RobWin ,This code has `isPermissionAcquired` flag which tells whether permission was acquired before or not. This is for completeness, trying to cover scenario where permission is not acquired before. However, we don't have a use case in the implementation yet.

Reply:
Since there is not hook like `whenComplete`, we stop the time when `future.get` is invoked. I think that's fine. We can document it in the JavaDoc.